src/IO/S3/Credentials.cpp

```diff
@@ -233,10 +224,81 @@ std::shared_ptr<AWSEC2MetadataClient> InitEC2MetadataClient(const Aws::Client::C
             }
         }
     }
-    LOG_INFO(logger, "Using IMDS endpoint: {}", ec2_metadata_service_endpoint);
-    return std::make_shared<AWSEC2MetadataClient>(client_configuration, ec2_metadata_service_endpoint.c_str());
+    return ec2_metadata_service_endpoint;
+}
+
+std::shared_ptr<AWSEC2MetadataClient> InitEC2MetadataClient(const Aws::Client::ClientConfiguration & client_configuration)
+{
+    auto endpoint = getAWSMetadataEndpoint();
+    return std::make_shared<AWSEC2MetadataClient>(client_configuration, endpoint.c_str());
 }
 
+String AWSEC2MetadataClient::getAvailabilityZoneOrException()
+{
+    Poco::URI uri(getAWSMetadataEndpoint() + EC2_AVAILABILITY_ZONE_RESOURCE);
+    Poco::Net::HTTPClientSession session(uri.getHost(), uri.getPort());
+    session.setTimeout(Poco::Timespan(AVAILABILITY_ZONE_REQUEST_TIMEOUT_SECONDS, 0));
+
+    Poco::Net::HTTPResponse response;
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, uri.getPath());
+    session.sendRequest(request);
+
+    std::istream & rs = session.receiveResponse(response);
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
+        throw DB::Exception(ErrorCodes::AWS_ERROR, "Failed to get AWS availability zone. HTTP response code: {}", response.getStatus());
+    String response_data;
+    Poco::StreamCopier::copyToString(rs, response_data);
+    return response_data;
+}
+
+String getGCPAvailabilityZoneOrException()
+{
+    Poco::URI uri(String(GCP_METADATA_SERVICE_ENDPOINT) + "/computeMetadata/v1/instance/zone");
+    Poco::Net::HTTPClientSession session(uri.getHost(), uri.getPort());
+    session.setTimeout(Poco::Timespan(AVAILABILITY_ZONE_REQUEST_TIMEOUT_SECONDS, 0));
+
+    Poco::Net::HTTPRequest request(Poco::Net::HTTPRequest::HTTP_GET, uri.getPath());
+    Poco::Net::HTTPResponse response;
+    request.set("Metadata-Flavor", "Google");
+
+    session.sendRequest(request);
+    std::istream & rs = session.receiveResponse(response);
+    if (response.getStatus() != Poco::Net::HTTPResponse::HTTP_OK)
+        throw DB::Exception(ErrorCodes::GCP_ERROR, "Failed to get GCP availability zone. HTTP response code: {}", response.getStatus());
+    String response_data;
+    Poco::StreamCopier::copyToString(rs, response_data);
+    Strings zone_info;
+    boost::split(zone_info, response_data, boost::is_any_of("/"));
+    /// We expect GCP returns a string as "projects/123456789/zones/us-central1a".
+    if (zone_info.size() != 4)
+        throw DB::Exception(ErrorCodes::GCP_ERROR, "Invalid format of GCP zone information, expect projects/<project-number>/zones/<zone-value>");
+    return zone_info[3];
+}
+
+String getRunningAvailabilityZone()
+{
+    LOG_INFO(&Poco::Logger::get("Application"), "Trying to detect the availability zone.");
+    try
+    {
+        auto aws_az = AWSEC2MetadataClient::getAvailabilityZoneOrException();
+        return aws_az;
+    }
+    catch (const std::exception & aws_ex)
```

Comment:
```suggestion
    catch (...)
```
and store a message with something like
```cpp
auto aws_ex_msg = getExceptionMessage(std::current_exception());
```
same for GCP below

Reply:
done.